In [1]:
import pandas as pd
import numpy as np

df = pd.read_json('https://raw.githubusercontent.com/onaio/ona-tech/master/data/water_points.json')

pd.set_option('display.max_columns', None)
df.head()

water_pay respondent research_asst_name water_used_season  \
0        no  community    Haruna Mohammed        year_round   
1        no  community     Haruna Mohmmed        year_round   
2        no  community    Haruna Mohammed        year_round   
3        no  community    Haruna Mohammed        year_round   
4        no  community    Haruna Mohammed        year_round   

  _bamboo_dataset_id _deleted_at water_point_condition     _xform_id_string  \
0                            NaT           functioning  _08_Water_points_CV   
1                            NaT           functioning  _08_Water_points_CV   
2                            NaT           functioning  _08_Water_points_CV   
3                            NaT           functioning  _08_Water_points_CV   
4                            NaT           functioning  _08_Water_points_CV   

  other_point_1km                                 _attachments  \
0              no  [north_ghana/attachments/1351696546452.jpg]   
1             yes  [north_ghana/attachments/1351701849971.jpg]   
2             yes  [north_ghana/attachments/1351702462336.jpg]   
3             yes  [north_ghana/attachments/1351702971561.jpg]   
4             yes  [north_ghana/attachments/1351703622326.jpg]   

  communities_villages                       end animal_number water_point_id  \
0            Gumaryili  2012-11-12T11:46:32.454Z      more_500            xxx   
1            Selinvoya  2012-11-12T11:49:36.619Z     50_to_500            xxx   
2            Selinvoya  2012-10-31T16:57:37.864Z     50_to_500            xxx   
3            Selinvoya  2012-10-31T17:06:55.047Z     50_to_500            xxx   
4            Selinvoya  2012-10-31T17:15:57.847Z     50_to_500            xxx   

                      start water_connected water_manager_name  \
0  2012-10-31T15:11:04.618Z              no  community members   
1  2012-10-31T16:41:49.738Z              no       Amadu Salifu   
2  2012-10-31T16:52:02.601Z              no   Sulemana Abdulai   
3  2012-10-31T16:58:46.672Z              no    Haruna Mohammed   
4  2012-10-31T17:08:27.160Z              no          Sulemana    

             _status  enum_id_1 water_lift_mechanism districts_divisions  \
0  submitted_via_web          5                   no            northern   
1  submitted_via_web          5                  yes            northern   
2  submitted_via_web          5                  yes            northern   
3  submitted_via_web          5                  yes            northern   
4  submitted_via_web          5                  yes            northern   

                              _uuid                grid       date  \
0  f8bcee72d7a0400fb99ae11bbf804010  grid_further_500_m 2012-10-31   
1  c2f6b298955f47ab9f177bee1214141d  grid_further_500_m 2012-10-31   
2  6bc6d188611d47f6a666cfd1eaa33998  grid_further_500_m 2012-10-31   
3  4b28ac4cbba744d79ba4257f772f94d6  grid_further_500_m 2012-10-31   
4  7893ce5321804f229e533f36e90c9c6f  grid_further_500_m 2012-10-31   

                       formhub/uuid road_available water_functioning  \
0  4d41d54d134c4bfa9078571addd819b9             no               yes   
1  4d41d54d134c4bfa9078571addd819b9            yes               yes   
2  4d41d54d134c4bfa9078571addd819b9            yes               yes   
3  4d41d54d134c4bfa9078571addd819b9            yes               yes   
4  4d41d54d134c4bfa9078571addd819b9            yes               yes   

     _submission_time signal water_source_type                _geolocation  \
0 2012-11-13 07:13:57    low        dam_dugout   [10.1892764, -0.66410362]   
1 2012-11-13 07:14:04   high  unprotected_well  [10.28173052, -0.56901122]   
2 2012-11-13 07:14:07   high          borehole  [10.28169238, -0.56962993]   
3 2012-11-13 07:14:14   high          borehole  [10.28115661, -0.56918339]   
4 2012-11-13 07:14:22   high          borehole  [10.28044635, -0.56723556]   

   water_point_image                             water_point_geocode  \
0  135169654645

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 712 entries, 0 to 711
Data columns (total 48 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   water_pay                    712 non-null    object        
 1   respondent                   712 non-null    object        
 2   research_asst_name           712 non-null    object        
 3   water_used_season            712 non-null    object        
 4   _bamboo_dataset_id           712 non-null    object        
 5   _deleted_at                  0 non-null      datetime64[ns]
 6   water_point_condition        712 non-null    object        
 7   _xform_id_string             712 non-null    object        
 8   other_point_1km              712 non-null    object        
 9   _attachments                 712 non-null    object        
 10  communities_villages         712 non-null    object        
 11  end                          712 non-null    

In [3]:
df.isnull().sum()

water_pay                        0
respondent                       0
research_asst_name               0
water_used_season                0
_bamboo_dataset_id               0
_deleted_at                    712
water_point_condition            0
_xform_id_string                 0
other_point_1km                  0
_attachments                     0
communities_villages             0
end                              0
animal_number                  184
water_point_id                   0
start                            0
water_connected                  0
water_manager_name               4
_status                          0
enum_id_1                        0
water_lift_mechanism             0
districts_divisions              0
_uuid                            0
grid                             0
date                             0
formhub/uuid                     0
road_available                   0
water_functioning                0
_submission_time                 0
signal              

## Data Cleaning

In [4]:
#Selecting relevant columns.
df1 = df[['communities_villages', 'water_point_id', 'water_functioning', 'water_point_condition', 'water_source_type', 'water_point_geocode']].copy()

df1.head()

communities_villages water_point_id water_functioning water_point_condition  \
0            Gumaryili            xxx               yes           functioning   
1            Selinvoya            xxx               yes           functioning   
2            Selinvoya            xxx               yes           functioning   
3            Selinvoya            xxx               yes           functioning   
4            Selinvoya            xxx               yes           functioning   

  water_source_type                             water_point_geocode  
0        dam_dugout   10.1892764 -0.66410362 155.10000610351563 5.0  
1  unprotected_well  10.28173052 -0.56901122 201.89999389648438 5.0  
2          borehole  10.28169238 -0.56962993 202.60000610351563 5.0  
3          borehole   10.28115661 -0.56918339 199.6999969482422 5.0  
4          borehole   10.28044635 -0.56723556 208.6999969482422 5.0

In [5]:
df1.sample(15)

communities_villages water_point_id water_functioning  \
706               Gaadem            Xxx               yes   
202            Kpatarigu            Xxx               yes   
605            Kanbangsa            XXX               yes   
60                Nayoku            Xxx               yes   
317                Jaata            XXX               yes   
261                  Soo            Xxx                no   
571             Abanyeri            XXX               yes   
405                Guuta            XXX               yes   
432               Zundem            XXX               yes   
611               Bandem            XXX                no   
588               Chansa            XXX               yes   
193            Kpatarigu            Xxx               yes   
100             Gbaarigu             HM                no   
679              Sikabsa            Xxx               yes   
213            Kpatarigu            Xxx               yes   

    water_point_condition water_source_type  \
706           functioning          borehole   
202           functioning    protected_well   
605           functioning  unprotected_well   
60            functioning    protected_well   
317           functioning    protected_well   
261                broken          borehole   
571           functioning  unprotected_well   
405           functioning    protected_well   
432           functioning  unprotected_well   
611                broken  unprotected_well   
588           functioning  unprotected_well   
193     newly_constructed  protected_spring   
100                broken  unprotected_well   
679           functioning          borehole   
213           functioning  unprotected_well   

                                water_point_geocode  
706   10.58929312 -1.21979975 186.1999969482422 5.0  
202    10.02957245 -1.2425624 202.8000030517578 5.0  
605               10.56844719 -1.18565698 189.5 5.0  
60   10.37901261 -0.76577499 185.60000610351563 5.0  
317   10.79937639 -1.18165826 232.8000030517578 5.0  
261   10.19013403 -1.11075125 147.6999969482422 5.0  
571   10.59095108 -1.24047449 195.1999969482422 5.0  
405   10.66902321 -1.34471884 209.1999969482422 5.0  
432  10.67343872 -1.36617128 208.60000610351563 5.0  
611               10.66974365 -1.29267237 237.0 5.0  
588               10.57584174 -1.19176335 192.0 5.0  
193               10.03042546 -1.24181466 193.0 5.0  
100   10.61669213 -0.72005843 178.6999969482422 5.0  
679    10.51265594 -1.3783035 192.3000030517578 5.0  
213  10.03377118 -1.24406943 198.89999389648438 5.0

In [6]:
#Checking counts of unique values for each column.
for col in df1.columns:
    print(col, df1[col].nunique(), len(df1))

communities_villages 65 712
water_point_id 81 712
water_functioning 3 712
water_point_condition 6 712
water_source_type 9 712
water_point_geocode 708 712


In [7]:
#Checking for null values.
df1.isnull().sum()

communities_villages     0
water_point_id           0
water_functioning        0
water_point_condition    0
water_source_type        0
water_point_geocode      3
dtype: int64

In [8]:
#Viewing missing data
df1[df1['water_point_geocode'].isnull()]

communities_villages water_point_id water_functioning  \
113            Namgurima             Sl                no   
114            Namgurima             HM                no   
674                Piisa            Xxx               yes   

    water_point_condition water_source_type water_point_geocode  
113             abandoned          borehole                 NaN  
114                broken          borehole                 NaN  
674           functioning          borehole                 NaN

In [9]:
#Replacing the missing water_point_geocode with their unique water_point_ids
df1['water_point_geocode'].fillna(df1.water_point_id, inplace=True)

df1['water_point_geocode'].isnull().sum()

0

In [10]:
#Checking for duplicates
df1.duplicated().sum()

1

In [11]:
df1[df1.duplicated()]

communities_villages water_point_id water_functioning  \
709              Vundema            Xxx               yes   

    water_point_condition water_source_type  \
709                broken          borehole   

                              water_point_geocode  
709  10.5379067 -1.38202276 229.6999969482422 5.0

In [12]:
len(df1)

712

In [13]:
#Dropping the duplicate.
df1.drop_duplicates(inplace=True)

print(len(df1))
type(df1)

711


pandas.core.frame.DataFrame

## Data Analysis

**How many water points are functional?**

In [14]:
df1.water_functioning.value_counts()

yes      622
no        87
na_dn      2
Name: water_functioning, dtype: int64

**ANS = There are 622 functional water points.**

**Calculating the number of water points per community:**

In [15]:
#Grouping the communities and counting the unique water points in each using water_point_geocode.
pd.set_option('display.max_rows', None)
number_water_points = df1.groupby('communities_villages')['water_point_geocode'].count().reset_index(name = 'waterpoints_count')

print(number_water_points)

   communities_villages  waterpoints_count
0              Abanyeri                  4
1           Akpari-yeri                  3
2              Alavanyo                  3
3                 Arigu                 12
4               Badomsa                 27
5                Bandem                  7
6             Banyangsa                 10
7              Bechinsa                 26
8           Chanpolinsa                  4
9                Chansa                  9
10             Chondema                  4
11               Dibisi                  2
12              Dorinsa                 17
13                Fiisa                  5
14               Gaadem                  2
15               Garigu                  1
16             Gbaarigu                  5
17                Gbima                  3
18            Gumaryili                  1
19                Guuta                 32
20           Guuta-Nasa                 11
21                Jaata                  8
22         

In [16]:
number_water_points['waterpoints_count'].sum()

711

**What is the rank for each community by the percentage of broken water points?**

In [17]:
#Unique values of water_point condition.
df1.water_point_condition.value_counts()

functioning           590
broken                 39
abandoned              36
newly_constructed      28
under_construction     16
na_dn                   2
Name: water_point_condition, dtype: int64

In [18]:
#Grouping the dataframe by communities_villages and water_point_condition to find the counts of each condition.
df2 = df1.groupby(['communities_villages', 'water_point_condition'])['water_point_geocode'].count().reset_index(name='count')

df2.head(15)

communities_villages water_point_condition  count
0              Abanyeri           functioning      3
1              Abanyeri     newly_constructed      1
2           Akpari-yeri           functioning      3
3              Alavanyo             abandoned      1
4              Alavanyo           functioning      2
5                 Arigu             abandoned      4
6                 Arigu                broken      2
7                 Arigu           functioning      6
8               Badomsa           functioning     26
9               Badomsa     newly_constructed      1
10               Bandem                broken      2
11               Bandem           functioning      4
12               Bandem     newly_constructed      1
13            Banyangsa                broken      1
14            Banyangsa           functioning      8

In [19]:
#Reorganizing the dataframe into a more useful format using pivoting.
pivoted = df2.pivot_table(index='communities_villages', columns='water_point_condition', values='count', fill_value=0).reset_index()

pivoted.head()

water_point_condition communities_villages  abandoned  broken  functioning  \
0                                 Abanyeri          0       0            3   
1                              Akpari-yeri          0       0            3   
2                                 Alavanyo          1       0            2   
3                                    Arigu          4       2            6   
4                                  Badomsa          0       0           26   

water_point_condition  na_dn  newly_constructed  under_construction  
0                          0                  1                   0  
1                          0                  0                   0  
2                          0                  0                   0  
3                          0                  0                   0  
4                          0                  1                   0

In [20]:
#Merging the two dataframes: number_water_points and pivoted in order to calculate rank.
new_df = pd.merge(number_water_points, pivoted)

new_df.head(10)

communities_villages  waterpoints_count  abandoned  broken  functioning  \
0             Abanyeri                  4          0       0            3   
1          Akpari-yeri                  3          0       0            3   
2             Alavanyo                  3          1       0            2   
3                Arigu                 12          4       2            6   
4              Badomsa                 27          0       0           26   
5               Bandem                  7          0       2            4   
6            Banyangsa                 10          0       1            8   
7             Bechinsa                 26          0       0           26   
8          Chanpolinsa                  4          0       0            4   
9               Chansa                  9          0       0            9   

   na_dn  newly_constructed  under_construction  
0      0                  1                   0  
1      0                  0                   0  
2      0                  0                   0  
3      0                  0                   0  
4      0                  1                   0  
5      0                  1                   0  
6      1                  0                   0  
7      0                  0                   0  
8      0                  0                   0  
9      0                  0                   0

In [21]:
#Calculating the community ranks, correct to 2 decimal places.
new_df['community_rank'] = new_df.apply(lambda row: (row['broken'] / row['waterpoints_count'])*100, axis=1).round(2)

new_df.head()

communities_villages  waterpoints_count  abandoned  broken  functioning  \
0             Abanyeri                  4          0       0            3   
1          Akpari-yeri                  3          0       0            3   
2             Alavanyo                  3          1       0            2   
3                Arigu                 12          4       2            6   
4              Badomsa                 27          0       0           26   

   na_dn  newly_constructed  under_construction  community_rank  
0      0                  1                   0            0.00  
1      0                  0                   0            0.00  
2      0                  0                   0            0.00  
3      0                  0                   0           16.67  
4      0                  1                   0            0.00

**ANS:**

In [22]:
#Sorting and displaying final results.
pd.set_option('display.max_rows', None)
result = new_df[['communities_villages', 'community_rank']].sort_values(by=['community_rank'])

print(result)

   communities_villages  community_rank
0              Abanyeri            0.00
28              Kalaasa            0.00
29            Kanbangsa            0.00
31                  Kom            0.00
33            Kpikpaluk            0.00
35              Kulbugu            0.00
36              Kunkwah            0.00
37               Kurugu            0.00
39            Logvasgsa            0.00
42           Mwalorinsa            0.00
43             Nabulugu            0.00
45              Nawaasa            0.00
47             Nyandema            0.00
49                Piisa            0.00
50            Selinvoya            0.00
51              Sikabsa            0.00
53                 Suik            0.00
54            Tankangsa            0.00
56                Tuisa            0.00
58           Zangu-Vuga            0.00
61                  Zua            0.00
62              Zuedema            0.00
26           Jiriwiensa            0.00
25     Jiningsa-Yipaala            0.00
